app/src/main/java/com/keylesspalace/tusky/viewmodel/EditProfileViewModel.kt

```diff
@@ -96,68 +105,49 @@ class EditProfileViewModel @Inject constructor(
         headerData.value = getHeaderUri()
     }
 
-    fun save(newDisplayName: String, newNote: String, newLocked: Boolean, newFields: List<StringField>) {
+    internal fun save(newProfileData: ProfileData) {
         if (saveData.value is Loading || profileData.value !is Success) {
             return
         }
 
         saveData.value = Loading()
 
-        val displayName = if (oldProfileData?.displayName == newDisplayName) {
-            null
-        } else {
-            newDisplayName.toRequestBody(MultipartBody.FORM)
-        }
-
-        val note = if (oldProfileData?.source?.note == newNote) {
-            null
-        } else {
-            newNote.toRequestBody(MultipartBody.FORM)
-        }
-
-        val locked = if (oldProfileData?.locked == newLocked) {
-            null
-        } else {
-            newLocked.toString().toRequestBody(MultipartBody.FORM)
-        }
-
-        val avatar = if (avatarData.value != null) {
-            val avatarBody = getCacheFileForName(AVATAR_FILE_NAME).asRequestBody("image/png".toMediaTypeOrNull())
-            MultipartBody.Part.createFormData("avatar", randomAlphanumericString(12), avatarBody)
-        } else {
-            null
-        }
-
-        val header = if (headerData.value != null) {
-            val headerBody = getCacheFileForName(HEADER_FILE_NAME).asRequestBody("image/png".toMediaTypeOrNull())
-            MultipartBody.Part.createFormData("header", randomAlphanumericString(12), headerBody)
-        } else {
-            null
-        }
-
-        // when one field changed, all have to be sent or they unchanged ones would get overridden
-        val fieldsUnchanged = oldProfileData?.source?.fields == newFields
-        val field1 = calculateFieldToUpdate(newFields.getOrNull(0), fieldsUnchanged)
-        val field2 = calculateFieldToUpdate(newFields.getOrNull(1), fieldsUnchanged)
-        val field3 = calculateFieldToUpdate(newFields.getOrNull(2), fieldsUnchanged)
-        val field4 = calculateFieldToUpdate(newFields.getOrNull(3), fieldsUnchanged)
-
-        if (displayName == null && note == null && locked == null && avatar == null && header == null &&
-            field1 == null && field2 == null && field3 == null && field4 == null
-        ) {
-            /** if nothing has changed, there is no need to make a network request */
+        val diff = getProfileDiff(apiProfileAccount, newProfileData)
+        if (!diff.hasChanges()) {
+            // if nothing has changed, there is no need to make an api call
             saveData.postValue(Success())
```

Comment:
Is postValue() the same as value = ?

(Should probably be consistent.)

Reply:
They aren't. Post Value does so by posting the task to a Main Thread if I correctly recall. 

So if you do 

`x.postValue(1) 
x.postValue(2)
x.value = 3
`

The value will be immediately set to 3, then to 2 (as it would overwrite the 1) when the "Runnable" is dispatched and run by the main Thread message queue. (it really uses a Runnable behind the scenes if I am not mistaken).

In my experience it's often safe(r) to use postValue but I guess it depends on the use Case. I haven't used LiveData in a bit (having switched most of the time to StateFlow/SharedFLow).

Reply:
According to the documentation (of those methods): It's only a question of where you call it as `setValue` needs to be done on the main thread.

The async character of postValue (overwriting stuff) is then probably only a secondary problem.

So this can be resolved.